src/wp-includes/canonical.php

```diff
@@ -543,6 +543,10 @@ function redirect_canonical( $requested_url = null, $do_redirect = true ) {
 		}
 	}
 
+	if ( is_attachment() && '1' !== get_option( 'wp_media_use_attachment_pages' ) ) {
+		$redirect_url = wp_get_attachment_url( get_query_var( 'attachment_id' ) );
+	}
+
```

Comment:
The redirect may need to consider the (front-end) visibility of the attachment post and eventually return a 404 or an appropriate error page if the current user is not allowed to see it.